language-plutus-core/src/Language/PlutusCore/Type.hs

```diff
@@ -91,29 +93,62 @@ instance Corecursive (Type tyname a) where
     embed (TyLamF l tn k ty)    = TyLam l tn k ty
     embed (TyAppF l ty ty')     = TyApp l ty ty'
 
--- | Substitute @a@ for @b@ in @c@.
-tyNameSubstitute :: Eq (tyname a)
-                 => tyname a -- ^ @a@
-                 -> tyname a -- ^ @b@
-                 -> Type tyname a -- ^ @c@
-                 -> Type tyname a
-tyNameSubstitute tn tn' = cata a where
-    a (TyVarF x tn'') | tn == tn'' = TyVar x tn'
-    a x               = embed x
-
-instance (Eq (tyname a), Eq a) => Eq (Type tyname a) where
-    (==) (TyVar a1 tn) (TyVar a2 tn')                   = a1 == a2 && tn == tn'
-    (==) (TyFun a1 ty ty') (TyFun a2 ty'' ty''')        = a1 == a2 && ty == ty'' && ty' == ty'''
-    (==) (TyFix a1 tn ty) (TyFix a2 tn' ty')            =
-        a1 == a2 && tyNameSubstitute tn' tn ty' == ty
-    (==) (TyForall a1 tn k ty) (TyForall a2 tn' k' ty') =
-        a1 == a2 && tyNameSubstitute tn' tn ty' == ty && k == k'
-    (==) (TyBuiltin a1 b) (TyBuiltin a2 b')             = a1 == a2 && b == b'
-    (==) (TyInt a1 n) (TyInt a2 n')                     = a1 == a2 && n == n'
-    (==) (TyLam a1 tn k ty) (TyLam a2 tn' k' ty')       =
-        a1 == a2 && tyNameSubstitute tn' tn ty' == ty && k == k'
-    (==) (TyApp a1 ty ty') (TyApp a2 ty'' ty''')        = a1 == a2 && ty == ty'' && ty' == ty'''
-    (==) _ _                                            = False
+-- this type is used for replacing type names in
+-- the Eq instance
+type EqState tyname a = M.Map (tyname a) (tyname a)
+
+rebind :: (Ord (tyname a), MonadState (EqState tyname a) m)
+       => tyname a
+       -> tyname a
+       -> m ()
+rebind = modify .* M.insert
+
+remove :: (Ord (tyname a), MonadState (EqState tyname a) m)
+       => tyname a
+       -> m ()
+remove = modify . M.delete
+
+rebindAndEq :: (Eq a, Ord (tyname a), MonadState (EqState tyname a) m)
+            => Type tyname a
+            -> Type tyname a
+            -> tyname a
+            -> tyname a
+            -> m Bool
+rebindAndEq ty ty' tn tn' =
+    rebind tn' tn *>
+    eqTypeM ty ty' <* remove tn'
+
+-- This tests for equality of names inside a monad that allows substitution.
+eqTypeM :: (Ord (tyname a), MonadState (EqState tyname a) m, Eq a)
+        => Type tyname a
+        -> Type tyname a
+        -> m Bool
+
+eqTypeM (TyFun _ ty ty') (TyFun _ ty'' ty''') = eqTypeM ty ty'' <&&> eqTypeM ty' ty'''
+eqTypeM (TyApp _ f a) (TyApp _ f' a') = eqTypeM f f' <&&> eqTypeM a a'
+
+eqTypeM (TyInt _ n) (TyInt _ n')              = pure (n == n')
+eqTypeM (TyBuiltin _ b) (TyBuiltin _ b')      = pure (b == b')
+
+eqTypeM (TyFix _ tn ty) (TyFix _ tn' ty') =
+    rebindAndEq ty ty' tn tn'
+eqTypeM (TyForall _ tn k ty) (TyForall _ tn' k' ty') = do
+    tyEq <- rebindAndEq ty ty' tn tn'
+    pure (k == k' && tyEq)
+eqTypeM (TyLam _ tn k ty) (TyLam _ tn' k' ty') = do
+    tyEq <- rebindAndEq ty ty' tn tn'
+    pure (k == k' && tyEq)
+
+eqTypeM (TyVar _ tn) (TyVar _ tn') = do
```

Comment:
Is that case correct? We could have just one argument being a variable that when looked up is a type equal to the other one. (Even in the variable case, both variables may be bound to other variables, which when looked up turn out to be the same.) Am I overlooking something?

Reply:
I think you're correct. I wrote above:

> Do we also need those recursive lookups that we had in the type checker?

Can we instead resolve this at insertion time (as opposed to lookup time) and store direct renamings?

Reply:
I think, the choice of whether to resolve at insertion or lookup time is one of style. Arguably looking up at insertion time is slightly more efficient when there are multiple look ups chasing the same chain (but less efficient if a variable is not looked up at all). On the other hand, chasing at look up time probably leads to simpler code as we usually do not look up during insertion at all, but if we want to collapse these chains, we also need look ups during insertion.

I don't care much either way as long as we are consistent about it and document it well (i.e., what is the invariant of the environment carried around).

Reply:
Since our current approach involves dropping type names when they go out of scope, I think resolving at lookup is safest.

Reply:
Thinking about this some more (and having triggered an infinite loop while performing recursive lookups), I do not believe it matters that we do recursive lookups. 

If we have two types, `tyLeft = (all a (type) a)` and `tyRight = (all b (type) b)`, we only really care to do one level of lookups, since we use the map to store lookups of right-names in terms of left-names (that is, the left-names are the keys), which are fundamentally different.

Reply:
@mchakravarty @effectfully Can I merge this? If the explanation was unclear I can try to be more precise, but I'm fairly certain that it is in fact correct.

Reply:
I'll take a close look tomorrow.

Reply:
I'm not confident, but I feel like Vanessa is right. Names can't reference previous names, they can only shadow them, but that's handled by the algorithm. Hence I do not feel like recursive lookups are necessary. I'm resolving this, but I'll also think more just maybe to come up with a better argument or disprove myself.